File: server/internal/transactions/categorizer/ollama.go
package categorizer

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strings"
	"time"
)

// ollamaBatchSize is the number of transactions classified per Ollama call —
// small, because local models degrade on long classification lists.
const ollamaBatchSize = 5

// OllamaCategorizer uses a local Ollama instance to classify transactions
// that weren't matched by rules or Plaid category mapping.
type OllamaCategorizer struct {
	baseURL    string
	model      string
	httpClient *http.Client
	categories []CategoryRef
	log        *slog.Logger
}

func NewOllamaCategorizer(baseURL, model string, categories []CategoryRef, log *slog.Logger) *OllamaCategorizer {
	return &OllamaCategorizer{
		baseURL:    strings.TrimRight(baseURL, "/"),
		model:      model,
		httpClient: &http.Client{Timeout: 300 * time.Second},
		categories: categories,
		log:        log,
	}
}

func (lc *OllamaCategorizer) BatchSize() int { return ollamaBatchSize }

// CategorizeBatch classifies multiple transactions in a single LLM call.
// globalExamples is a slice of top-merchant historical examples injected into
// every prompt batch as a few-shot reference. Returns only successfully
// classified results; unparseable or low-confidence entries are dropped.
func (lc *OllamaCategorizer) CategorizeBatch(ctx context.Context, txns []TransactionInput, globalExamples []ExampleTransaction) ([]LLMResult, error) {
	if len(txns) == 0 {
		return nil, nil
	}

	prompt := buildPrompt(lc.categories, txns, globalExamples)
	resp, err := lc.callOllama(ctx, prompt)
	if err != nil {
		return nil, err
	}
	return parseResponse(resp, txns, lc.categories, lc.log), nil
}

type ollamaRequest struct {
	Model  string `json:"model"`
	Prompt string `json:"prompt"`
	Stream bool   `json:"stream"`
	Format string `json:"format,omitempty"`
	// Think is sent as a literal false (no omitempty) so reasoning models
	// (qwen3, deepseek-r1, …) skip their <think> block and answer directly —
	// essential for fast, structured categorization on CPU. Non-reasoning
	// models (qwen2.5, llama3.1) accept think:false as a no-op.
	Think   bool           `json:"think"`
	Options map[string]any `json:"options,omitempty"`
}

type ollamaResponse struct {
	Response string `json:"response"`
	Done     bool   `json:"done"`
}

func (lc *OllamaCategorizer) callOllama(ctx context.Context, prompt string) (string, error) {
	reqBody := ollamaRequest{
		Model:  lc.model,
		Prompt: prompt,
		Stream: false,
		Format: "json",
		Think:  false,
		Options: map[string]any{
			"temperature": 0.1,
			"num_predict": 2048,
		},
	}

	body, err := json.Marshal(reqBody)
	if err != nil {
		return "", fmt.Errorf("marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, lc.baseURL+"/api/generate", bytes.NewReader(body))
	if err != nil {
		return "", fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := lc.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("ollama http: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		respBody, err := io.ReadAll(resp.Body)
		if err != nil {
			return "", fmt.Errorf("read ollama error response: %w", err)
		}
		return "", fmt.Errorf("ollama returned %d: %s", resp.StatusCode, string(respBody))
	}

	var ollamaResp ollamaResponse
	if err := json.NewDecoder(resp.Body).Decode(&ollamaResp); err != nil {
		return "", fmt.Errorf("decode ollama response: %w", err)
	}

	return ollamaResp.Response, nil
}

File: server-rs/src/clients/ollama.rs
use anyhow::{Context, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};

const TIMEOUT_SECONDS: u64 = 300;

#[derive(Clone)]
pub struct OllamaClient {
    base_url: String,
    model: String,
    http: Client,
}

impl OllamaClient {
    pub fn new(base_url: impl Into<String>, model: impl Into<String>) -> Result<Self> {
        Ok(Self {
            base_url: base_url.into().trim_end_matches('/').to_owned(),
            model: model.into(),
            http: Client::builder()
                .timeout(std::time::Duration::from_secs(TIMEOUT_SECONDS))
                .build()
                .context("build Ollama HTTP client")?,
        })
    }

    pub async fn generate(&self, prompt: &str) -> Result<String> {
        let response = self
            .http
            .post(format!("{}/api/generate", self.base_url))
            .json(&GenerateRequest {
                model: &self.model,
                prompt,
                stream: false,
                format: "json",
                // Think is sent as a literal false (no omitempty) so reasoning models
                // (qwen3, deepseek-r1, …) skip their <think> block and answer directly —
                // essential for fast, structured categorization on CPU. Non-reasoning
                // models (qwen2.5, llama3.1) accept think:false as a no-op.
                think: false,
                options: GenerateOptions {
                    temperature: 0.1,
                    num_predict: 2048,
                },
            })
            .send()
            .await
            .context("ollama http")?;
        let status = response.status();
        if status != reqwest::StatusCode::OK {
            let body = response.text().await.context("read Ollama error response")?;
            anyhow::bail!("ollama returned {}: {body}", status.as_u16());
        }
        let body = response.text().await.context("read Ollama response")?;
        Ok(serde_json::from_str::<GenerateResponse>(&body)
            .context("decode Ollama response")?
            .response)
    }
}

#[derive(Serialize)]
struct GenerateRequest<'a> {
    model: &'a str,
    prompt: &'a str,
    stream: bool,
    format: &'static str,
    think: bool,
    options: GenerateOptions,
}

#[derive(Serialize)]
struct GenerateOptions {
    temperature: f64,
    num_predict: u16,
}

#[derive(Deserialize)]
struct GenerateResponse {
    response: String,
}

#[cfg(test)]
mod tests {
    use serde_json::json;
    use wiremock::{
        Mock, MockServer, ResponseTemplate,
        matchers::{body_json, method, path},
    };

    use super::OllamaClient;

    #[tokio::test]
    async fn generates_json_without_reasoning() {
        let server = MockServer::start().await;
        Mock::given(method("POST"))
            .and(path("/api/generate"))
            .and(body_json(json!({
                "model": "qwen",
                "prompt": "categorize this",
                "stream": false,
                "format": "json",
                "think": false,
                "options": {"temperature": 0.1, "num_predict": 2048}
            })))
            .respond_with(ResponseTemplate::new(200).set_body_json(json!({"response": "{}", "done": true})))
            .mount(&server)
            .await;

        assert_eq!(
            OllamaClient::new(format!("{}/", server.uri()), "qwen")
                .unwrap()
                .generate("categorize this")
                .await
                .unwrap(),
            "{}"
        );
    }
}
